### PR TITLE
CAPX-37 | Organizations...

### PR DESCRIPTION
**Pull request by sherakama**
# Ready for review
- Autocomplete for organizations
- multiple organization allowance
- child org import checkbox
- organization sync task
- Updated settings page to include link to sync 
- added sync url
- added warning messages on those pages that should have them
- created batch process for update

### CODE REVIEW

**Comment by sherakama** [comment]
@pingers @mpriest 

Please review this pull request for quality and functionality. To test functionality navigate to either the settings page or an importer page. If you are on an importer page you should be prompted with a message saying the organization codes are not available and to sync now. Press the sync now button and a batch should run. The result will be a new taxonomy full of organization codes. Once you have the org codes go back to the importer form and test out the autocomplete for both org code lookup and org name lookup. Try adding a few orgs.

Thanks,

File: includes/CAPx/Drupal/Organizations/Orgs.php [hunk 1]
```diff
@@ -0,0 +1,233 @@
+<?php
+/**
+ * @file
+ * @author [author] <[email]>
+ */
+
+namespace CAPx\Drupal\Organizations;
+
+define('CAPxORGTAXONOMY', 'capx_organizations');
```
Comment (mpriest) [inline includes/CAPx/Drupal/Organizations/Orgs.php]: If this is never going to be used elsewhere, could be a constant.

File: includes/CAPx/Drupal/Organizations/Orgs.php [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+   */
+  public static function getOrganizations($parent = 0) {
+    $vocab = Orgs::getVocabulary();
+    return taxonomy_get_tree($vocab->vid, $parent);
+  }
+
+  /**
+   * Gets and syncs the Organization data from the CAP API
+   * @param  $batch  boolean value on wether or not to use batch api.
+   * @todo: This function can take a while and should be moved to batch.
```
Comment (mpriest) [inline includes/CAPx/Drupal/Organizations/Orgs.php]: Isn't it using batch now?

File: stanford_capx.forms.inc [hunk 3]
```diff
@@ -845,10 +852,18 @@ function stanford_capx_importer_form($form, &$form_state) {
...
     '#description' => t('Please select the name of the organization you wish to import. Learn more about !orgchart.', array('!orgchart' => $link)),
     '#default_value' => isset($importer->organization) ? $importer->organization : '',
-    '#autocomplete_path' => 'stanford-cap-api-org-autocomplete', // @todo: this.
+    '#autocomplete_path' => 'admin/config/capx/autocomplete/orgs', // @todo: this.
+  );
+
+  $form['groups']['child_orgs'] = array(
+    '#type' => 'checkbox',
+    '#title' => t('Include child organizations'),
+    '#default_value' => '',
```
Comment (sherakama) [inline stanford_capx.forms.inc]: Two defaults Shea? What are you thinking.

File: includes/CAPx/Drupal/Organizations/Orgs.php (hunk 1 above)
Comment (sherakama) [inline includes/CAPx/Drupal/Organizations/Orgs.php]: Yea, good point. I would rather go through the static functions for the vocab.

File: includes/CAPx/Drupal/Organizations/Orgs.php (hunk 2 above)
Comment (sherakama) [inline includes/CAPx/Drupal/Organizations/Orgs.php]: Ha, yea. 
Need to clean up my comments it looks like.

File: includes/CAPx/Drupal/Organizations/Orgs.php [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+    // Inform the batch engine that we are not finished,
+    // and provide an estimation of the completion level we reached.
+    if ($context['sandbox']['progress'] !== $context['sandbox']['max']) {
+      $context['finished'] = $context['sandbox']['progress'] / $context['sandbox']['max'];
+      $context['message'] = t('Now processing group ' . $context['sandbox']['progress'] . " of " . $context['sandbox']['max']);
+    }
+  }
+
+  /**
+   * Takes an Org tree array response from the server and
```
Comment (mpriest) [inline includes/CAPx/Drupal/Organizations/Orgs.php]: I think this warrants a comment that although taxonomy_term_save() supports automatically updating parents/children, the source data is not hierarchical, but alphabetical... which is why this isn't so trivial.

File: stanford_capx.forms.inc [hunk 5]
```diff
@@ -781,6 +781,13 @@ function stanford_capx_importer_form($form, &$form_state) {
 
   global $importer;
 
+  // Do a check to see if the organization codes are available. If not, prompt
+  // the user to update them.
+  $hasorgs = \CAPx\Drupal\Organizations\Orgs::checkOrgs();
```
Comment (mpriest) [inline stanford_capx.forms.inc]: Should be $has_orgs.

File: stanford_capx.module [hunk 6]
```diff
@@ -395,121 +433,78 @@ function stanford_capx_admin_config_import_edit($importer_machine_name) {
...
+ * Autocomplete callback endpoint for
+ * @param  [type] $type [description]
+ * @return [type]       [description]
  */
-function stanford_capx_debug() {
-  $output = "";
+function stanford_capx_autocomplete($type, $string) {
+  $matches = array();
 
-  // $importer_config = array();
```
Comment (mpriest) [inline stanford_capx.module:404]: Farewell debug code!

File: stanford_capx.module [hunk 7]
```diff
@@ -395,121 +433,78 @@ function stanford_capx_admin_config_import_edit($importer_machine_name) {
...
 
+  $vocab = Orgs::getVocabulary();
 
-  // $mapper = new EntityMapper($mapper_config);
-  // $importer = new EntityImporter($importer_config, $mapper, $client);
-  // $importer->execute();
+  // Partial code lookup. eg: AAH for AAHH
+  // Partial name lookup. eg: Busi for Business and Operations
+  $or = db_or();
+  $or->condition('name', db_like($last_item) . '%', 'LIKE');
```
Comment (mpriest) [inline stanford_capx.module]: This usually should look like

``` php
  db_or()
    ->condition(...)
    ->condition(...);
```

not a big deal.

File: stanford_capx.forms.inc (hunk 3 above)
Comment (mpriest) [inline stanford_capx.forms.inc]: :+1:

Comment (jbickar) [inline stanford_capx.forms.inc]: John Bickar likes this

**Pull request merged**